METRON-622: Create a Metron Docker Compose application

merrimanr opened the pull request:
This PR is an initial attempt at creating a Docker development and testing environment for Metron.  Information about what's included and how to use it can be found in the README.

cestella commented on metron-docker/storm/bin/init-storm.sh:
```diff
@@ -0,0 +1,19 @@
+#  limitations under the License.
+#
+METRON_VERSION=0.3.0
+cd /opt/hbase-1.1.6/conf && jar uf $METRON_HOME/lib/metron-enrichment-$METRON_VERSION-uber.jar hbase-site.xml
```
Aren't we doing this by modifying the storm classpath in the other environments (the management pack and ansible)?  We should probably have the same mechanism everywhere.  Also, we can call into hbase from the parser topology.

cestella replied:
Oh,also, do we need the `METRON_VERSION` since it's in the `.env`?

mmiklavc replied:
We no longer want to modify the uber jars. This should be set via Storm's param topology.classpath in storm-site. ie. topology.classpath = "/etc/hbase/conf:/etc/hadoop/conf"

See https://github.com/apache/incubator-metron/pull/204

merrimanr replied:
Done.

cestella commented on pom.xml:
```diff
@@ -263,6 +263,10 @@
                         <!-- Packer/Bento non-source directory -->
                         <exclude>**/packer_cache/**</exclude>
+                        <exclude>**/hbase/data/**</exclude>
+                        <exclude>**/hbase/bin/init-commands.txt</exclude>
```
Don't need this if we can add a license header.

cestella commented on metron-docker/kafkazk/bin/init-zk.sh:
```diff
@@ -0,0 +1,22 @@
+#
+echo "create /metron metron" | ./bin/zookeeper-shell.sh localhost:2181
+echo "create /metron/topology topology" | ./bin/zookeeper-shell.sh localhost:2181
+echo 'create /metron/topology/global {"solr.zookeeper":"metron-kafkazk:2181","solr.collection":"metron","solr.numShards":1,"solr.replicationFactor":1,"es.clustername":"elasticsearch","es.ip":"metron-elasticsearch","es.port":"9300","es.date.format":"yyyy.MM.dd.HH"}' | ./bin/zookeeper-shell.sh localhost:2181
```
I get the creations of the directories in zk, but why are we pushing configs here instead of using the zk_load_utils shell script to submit the ones we deploy?

merrimanr replied:
I removed the line that pushes global config.

cestella commented on metron-docker/README.md:
````diff
@@ -0,0 +1,68 @@
+docker-compose exec kafkazk bash
+```
+
+These service names can be found in the docker-compose.yml file.
````
Can you talk a bit about metron-docker/kafkazk/bin/produce-data.sh and show an example of its use in the README?

merrimanr replied:
Done.  Also improved the scripts to include a message frequency and gave a usage example in the README.

mmiklavc commented on metron-docker/install-metron.sh:
```diff
@@ -0,0 +1,39 @@
+if [ $# -gt 0 ] && [ $1 == "-b" ]
+    then cd $METRON_PLATFORM_ROOT && mvn clean package -DskipTests
+fi
+mkdir -p $METRON_DOCKER_ROOT/mysql/enrichment/
```
Are these directories in a gitignore somewhere? Seems like we'll end up with new untracked files when we run the install command.

merrimanr replied:
yes they are in metron-docker/.gitignore

mmiklavc commented:
Also, this is great - thanks for submitting this!

merrimanr commented:
Just pushed out a commit that attempts to address these comments.  Includes:
- hbase config is now read from Storm classpath instead of hbase jar
- removed global config push from kafkazk init script
- added license to hbase/bin/init-commands.txt
- improved data generator script and updated documentation

merrimanr commented on metron-docker/kafkazk/bin/wait-for-it.sh:
```diff
@@ -0,0 +1,161 @@
+#!/usr/bin/env bash
+#   Use this script to test if a given TCP host/port are available
```
For security reasons, Docker will not pull in any assets outside of the directory containing the Dockerfile.  You have to do that separately or provide a script.  

We do have a script that copies in Metron dependencies at metron-docker/install-metron.sh.  Should we move the wait-for-it.sh script up to metron-docker and add steps to copy to images in install-metron.sh?

mmiklavc replied:
It feels like we should have some type of packaging step in order to accomplish the following:
- Copy resources to final structure for the deploy
- Perform parameterization
- Eliminate file duplication

merrimanr replied:
I could create a pom.xml that essentially replaces the need to manually run metron-docker/install-metron.sh and will instead perform those steps anytime Metron is packaged with Maven.  Will that work?  It should also be forgiving if Docker is not installed and just skip to the next module if it isn't.

mmiklavc replied:
Something along those lines. Check out the rpm packaging pom.

https://github.com/apache/incubator-metron/blob/master/metron-deployment/packaging/docker/rpm-docker/pom.xml

merrimanr replied:
Done

merrimanr commented on metron-docker/hbase/bin/init-commands.txt:
```diff
@@ -0,0 +1,5 @@
+create 'access_tracker', 'cf'
```
Done.

merrimanr commented on metron-docker/hbase/bin/init-hbase.sh:
```diff
@@ -0,0 +1,21 @@
+#
+export HBASE_CLASSPATH=`${HBASE_HOME}/bin/hbase classpath`
+./bin/hbase shell ./bin/init-commands.txt
+java -cp $HBASE_CLASSPATH:/usr/metron/$METRON_VERSION/lib/metron-data-management-$METRON_VERSION.jar org.apache.metron.dataloads.nonbulk.flatfile.SimpleEnrichmentFlatFileLoader -e /conf/enrichment-extractor.json -t enrichment -c cf -i /data/enrichments.csv
```
I could not use the flatfile_loader.sh script because it depends on the hadoop cli.  This is not necessary to load data into HBase and isn't installed in the image.

merrimanr replied:
Done in the latest commit.

merrimanr commented on metron-docker/storm/bin/start.sh:
```diff
@@ -0,0 +1,19 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+cd /opt/hbase-1.1.6/conf && jar uf $METRON_HOME/lib/metron-enrichment-$METRON_VERSION-uber.jar hbase-site.xml
```
Done.

ottobackwards commented:
I have followed the instructions, and it looks like it all worked but I am not sure what to do next to test this out.  The README is missing a usage example or something similar to the output of the ansible install that lists the services and ports

merrimanr commented:
In my experience I've found that if you run these commands:

eval $(docker-machine env metron-machine)
docker ps -a

and all containers are reported as up:

CONTAINER ID        IMAGE               COMMAND                  CREATED
          STATUS              PORTS
                                                      NAMES
a2f28ac154f3        metron_storm        "/bin/bash /home/stor"   2 days ago
         Up 2 days           0.0.0.0:8000->8000/tcp,
0.0.0.0:8080-8081->8080-8081/tcp
                                           metron_storm_1
7f40995c1344        metron_hbase        "/bin/sh -c ./bin/sta"   2 days ago
         Up 2 days           8080/tcp, 8085/tcp, 9090/tcp, 9095/tcp,
16000/tcp, 16201/tcp, 16301/tcp, 0.0.0.0:16010->16010/tcp   metron_hbase_1
013a5b660f2d        metron_kibana       "/docker-entrypoint.s"   2 days ago
         Up 2 days           0.0.0.0:5601->5601/tcp
                                                     metron_kibana_1
cdb8a20d1f2d        metron_mysql        "docker-entrypoint.sh"   2 days ago
         Up 2 days           0.0.0.0:3306->3306/tcp
                                                     metron_mysql_1
32f28acd2421        elasticsearch:2.3   "/docker-entrypoint.s"   2 days ago
         Up 2 days           0.0.0.0:9200->9200/tcp, 0.0.0.0:9300->9300/tcp
                                                     metron_elasticsearch_1
6a8aaa5c3853        metron_kafkazk      "/bin/sh -c ./bin/sta"   2 days ago
         Up 2 days           0.0.0.0:2181->2181/tcp, 0.0.0.0:9092->9092/tcp
                                                     metron_kafkazk_1

It's a pretty good indication that everything is working.  If 1 or more
containers have a status of "Exited" that means something went wrong.

There are some UIs that are included as well and I agree we should add
those urls to the README.  Will get that pushed out soon.

Ryan


On Wed, Dec 14, 2016 at 3:53 PM, ottobackwards <notifications@github.com>
wrote:

> I have followed the instructions, and it looks like it all worked but I am
> not sure what to do next to test this out. The README is missing a usage
> example or something similar to the output of the ansible install that
> lists the services and ports
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-metron/pull/393#issuecomment-267168563>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AATCeCXJW9x2lHv8hwP7sGFl10vh66w2ks5rIGVYgaJpZM4LJZHT>
> .
>

cestella commented:
I suspect we should ensure that data is flowing through.  @merrimanr , maybe it'd be good to document in the `README.md` a worked example where you:
* spin up a docker cluster
* generate data using the data generation scripts
* look at the indices to see that there are data
* check Kibana for the data

Am I missing anything?

ottobackwards commented:
That is a start @cestella @merrimanr.  Enough to validate the PR.  For the intended use case - we should have as much documentation for this as the vagrant stuff.  So wiki peer entries etc.  At least I would think.

ottobackwards commented:
I will continue to validate this when that next push goes out.

kylerichardson commented:
@merrimanr I'm super excited about this one! Thanks for your work!

While kicking the tires on it, I did notice something that isn't really a bug but got me thinking... The README assumes the developer is using a Mac. I personally use Fedora for most of my development and testing of Metron.

In my case, I found that without using docker-machine, the kafkazk service did not start up properly due to the use of sed on the DOCKER_HOST variable in the Dockerfile to set the advertised.listeners parameter in kafka server.properties. I was using the docker-engine on the localhost and thought I would avoid the hassle of docker-machine. It was easy enough to solve by manually manipulating the server.properties file in the container, but I was wondering if this could be avoided somehow?

I assume that had I installed docker-machine and used the provided script it would have worked as expected even on Fedora (I'm planning to test that case as well), but was curious why the use of the virtualbox driver? Is this a Mac specific requirement?

Thanks again for the contribution! I know this will make coding/testing new parsers a lot faster.

merrimanr commented:
Just pushed out changes based on this initial round of feedback.  Main changes were moving the install steps to a Maven pom and adding more detail to the documentation.

merrimanr commented:
@kylerichardson thank you for reviewing it!  I updated the documentation to include Docker for Mac or Docker for Windows.  One of the primary requirements for the way I use this is that containers must be accessible from my local environment where my IDE is running so services that broadcast their host address are tricky.  The Kafka advertised listener thing was the single most challenging issue I faced so not surprised you hit it too.  The Kafka Dockerfile is wired to pull the DOCKER_HOST from an input argument and the compose file is wired to pass the local DOCKER_HOST environment variable as the DOCKER_HOST input argument to the Kafka Dockerfile.  So you should be able to set your local DOCKER_HOST environment variable as such:

$ export DOCKER_HOST="tcp://<your docker host>:2376"

When you run "eval $(docker-machine env metron-machine)" that's pretty much what Docker Machine is doing, setting local environment variables to match the desired host.  Then after you build the environment, the advertised.listener property should be set to <your docker host> without you having to manually change it.  

I only used the virtualbox drive because that's what came out of the box.

kylerichardson commented:
Thanks for the explanation @merrimanr. I totally agree on the need for the
local IDE to have access to the containerized services. It's an easy enough
fix for me to manipulate the DOCKER_HOST env variable on Fedora and avoid
using docker-machine (although, I do see that as a big win for ease of use
on Mac and Windows).

I haven't had much luck with vagrant myself so am super excited for a
docker alternative to quick dev. As an aside, there are some Ansible
modules for docker (
https://docs.ansible.com/ansible/list_of_cloud_modules.html#docker) that we
could look into for incorporating some of the setup scripts and image
building to make it even more user friendly. Might just be a low priority
nice to have but thought I'd throw it out there.

On Mon, Dec 19, 2016 at 4:32 PM, merrimanr <notifications@github.com> wrote:

> @kylerichardson <https://github.com/kylerichardson> thank you for
> reviewing it! I updated the documentation to include Docker for Mac or
> Docker for Windows. One of the primary requirements for the way I use this
> is that containers must be accessible from my local environment where my
> IDE is running so services that broadcast their host address are tricky.
> The Kafka advertised listener thing was the single most challenging issue I
> faced so not surprised you hit it too. The Kafka Dockerfile is wired to
> pull the DOCKER_HOST from an input argument and the compose file is wired
> to pass the local DOCKER_HOST environment variable as the DOCKER_HOST input
> argument to the Kafka Dockerfile. So you should be able to set your local
> DOCKER_HOST environment variable as such:
>
> $ export DOCKER_HOST="tcp://:2376"
>
> When you run "eval $(docker-machine env metron-machine)" that's pretty
> much what Docker Machine is doing, setting local environment variables to
> match the desired host. Then after you build the environment, the
> advertised.listener property should be set to without you having to
> manually change it.
>
> I only used the virtualbox drive because that's what came out of the box.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-metron/pull/393#issuecomment-268084862>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJ6R3a4JoffRcq9etOoUAJExjJAeue-aks5rJvf_gaJpZM4LJZHT>
> .
>

ottobackwards commented:
So just to be clear, the end result of this is the cluster deployed, but nothing installed?  No topologies, no indices in ES etc?

merrimanr commented:
I would consider the topologies installed, just not running.  But yes, no data flowing end to end by default.

ottobackwards commented:
If the indices are not installed, then the topologies are moot.  I am sorry, I don't mean to be difficult, I am just not sure if it is working and I don't understand what the result should be or if it did not work

merrimanr commented:
There were several examples added to the README.  The last one "Run sensor data end to end" will result in data flowing into an ES index if executed properly.  If you are able to successfully follow the examples then it works.  There is also a "Usage" section that explains how to check the container status and lists the UIs.

I would argue that the topologies are not moot just because indices are not installed.  What if I'm working on a new parser class?  I will more likely be inspecting the enrichments topic to verify the new parser is working correctly rather than looking at the indices.  That would come later.

Maybe the scope of the PR is different than what you are expecting?  This originally was intended to be a light-weight development environment that allows a developer to work on or troubleshoot specific parts of the Metron flow, not an end-to-end solution.  We already have something for that (vagrant quickdev).

ottobackwards commented:
ok, I was not using the updated examples in the read me to verify the pr.  It was not clear to me, after starting everything up, what I should expect to be working and what should not be working.  I will go back and do that get the context.

kylerichardson commented:
I've run this up and successfully tested it using the examples provided in the README. It works as documented on docker-machine/boot2docker. Nice job.

One showstopper for me. I can't seem to find the topology logs in the storm container. I checked /var/log/storm and no topology specific logs were ever written. For debugging new parsers, etc. this will be important to have available.

I also want to highlight a few nice-to-haves that I would be perfectly happy submitting as separate, follow-on PRs.
- Custom the elasticsearch image (1) to have the elasticsearch-head plugin installed as part of the image build, (2) copy the es_templates into the image so they are available, (3) for bonus points, deploy the templates on container start
- Load zookeeper config on container start
- Improve kafkazk Dockerfile for local docker-engine (Linux); current problem is that DOCKER_HOST in this case defaults to empty string and the default argument in the Dockerfile is never hit
- Add an HDFS container to allow for complete testing of the indexing topology

merrimanr commented:
The logs for this container are written to /usr/share/apache-storm/logs/ instead of /var/log/storm.  That's probably why you don't see them.  That's just the way they configured it in this Docker image but I'm sure it's easy enough to change.

You can also view the logs through the Storm UI.  The one catch is that once you drill into a specific log (will give you a 404 initially) you have to replace the ip address (it's an internal Docker container IP and not accessible outside of the Docker host) with the Docker machine ip.

I agree those are nice add ons.  This is a pretty significant PR as is so I tried to keep the scope as small as possible while still providing the ability to flow data end to end.  The first one isn't too bad and I think it fits nicely into a separate PR.  The 2nd one is trival, I'm fine with configs being loaded initially and will make that change.  The 3rd one might be easy, you have the environment setup so maybe you could experiment with it (I will help)?  The last one will definitely require some work and should be a separate PR.

ottobackwards commented:
Maybe we can create jiras for these, and someone else might help?

kylerichardson commented:
I'll created METRON-646 for the elasticsearch image customizations. I already have a start on those changes.

kylerichardson commented:
@merrimanr Thanks for pointing out the location of the storm logs. I think it would be worth adding that to the README.

I have a fix for the kafkazk image running on local docker-engine (using unix socket). It requires an additional ARG in the Dockerfile (and docker-compose.yml) and an extra sed statement when producing the advertised.listeners parameter.

docker-compose.yml
```
...
kafkazk:
  build:
    context: ./kafkazk
  args:
    DOCKER_HOST: $DOCKER_HOST
+   BROKER_IP_ADDR: $BROKER_IP_ADDR
    METRON_VERSION: $METRON_VERSION
...
```

kafkazk/Dockerfile
```
...
+ARG DOCKER_HOST
+ARG BROKER_IP_ADDR
ARG METRON_VERSION
...
RUN echo -n 'advertised.listeners=PLAINTEXT://' >> /opt/kafka_2.11-0.10.0.0/config/server.properties
+RUN echo $DOCKER_HOST | sed "s/^$/"$BROKER_IP_ADDR":/g" | sed "s/tcp:\\/\\///g" | sed "s/:.*/:9092/g" >> /opt/kafka_2.11-0.10.0.0/config/server.properties
...
```

I would also suggest a note in the documentation. In the Setup section of the README, maybe something along the lines of... If you wish to use a local docker-engine install, please set an environment variable BROKER_IP_ADDR to the IP address of your host machine. This cannot be the loopback address.

ottobackwards commented:
When trying to verify the Run Sensor Data End to end with two terminals I get :
ottofowler@Winterfell ~/src/apache/work/incubator-metron-pr-393/metron-docker/compose
1 pr-393 % docker-compose exec kafkazk ./bin/produce-data.sh /data/BroExampleOutput.txt bro 

WARNING: The DOCKER_HOST variable is not set. Defaulting to a blank string.
Emitting data in /data/BroExampleOutput.txt to Kafka topic bro every 1 second(s)
[2017-01-06 19:28:49,385] ERROR Error when sending message to topic bro with key: null, value: 482 bytes with error: (org.apache.kafka.clients.producer.internals.ErrorLoggingCallback)
org.apache.kafka.common.errors.TimeoutException: Failed to update metadata after 60000 ms.


and

ottofowler@Winterfell:metron-docker/compose ‹pr-393›$ docker-compose exec kafkazk ./bin/kafka-console-consumer.sh --zookeeper localhost:2181 --topic bro

WARNING: The DOCKER_HOST variable is not set. Defaulting to a blank string.
No brokers found in ZK.


                                                                                             

Is there something I'm doing wrong?

merrimanr commented:
@ottobackwards you need to run this before you execute any docker or docker-compose command:

eval "$(docker-machine env metron-machine)"

This is documented in the "Setup" section of the README.  Is there a change we can make to the documentation to make this clearer?

merrimanr commented:
Thanks @kylerichardson!  I will add your change in along with initially loading all configs over the weekend.

ottobackwards commented:
I have run that - I have been able to do the steps in the readme ( except the topology which i haven't tried ) up until that point ( so my ZK is configured etc )

ottofowler@Winterfell ~/src/apache/work/incubator-metron-pr-393/metron-docker/compose
0 pr-393 % eval "$(docker-machine env metron-machine)"
Error checking TLS connection: Host is not running

ottofowler@Winterfell ~/src/apache/work/incubator-metron-pr-393/metron-docker/compose
0 pr-393 % docker-compose exec kafkazk ./bin/produce-data.sh /data/BroExampleOutput.txt bro

WARNING: The DOCKER_HOST variable is not set. Defaulting to a blank string.
Emitting data in /data/BroExampleOutput.txt to Kafka topic bro every 1 second(s)
[2017-01-06 19:47:02,967] ERROR Error when sending message to topic bro with key: null, value: 482 bytes with error: (org.apache.kafka.clients.producer.internals.ErrorLoggingCallback)
org.apache.kafka.common.errors.TimeoutException: Failed to update metadata after 60000 ms.

ottobackwards commented:
Actually - the test output data stuff was failing too, I just hadn't waited long enough to see since I wasn't trying to read.

merrimanr commented:
"Error checking TLS connection: Host is not running" that means it didn't work.  Can you try running "docker-machine ls" and ensure you have a machine called "metron-machine"?  It should look like this:

HW10673:compose rmerriman$ docker-machine ls
NAME             ACTIVE   DRIVER       STATE     URL                         SWARM   DOCKER    ERRORS
metron-machine   *        virtualbox   Running   tcp://192.168.99.102:2376           v1.12.5
HW10673:compose rmerriman$ eval "$(docker-machine env metron-machine)"
HW10673:compose rmerriman$ echo $DOCKER_HOST
tcp://192.168.99.102:2376

ottobackwards commented:
ottofowler@Winterfell ~/src/apache/work/incubator-metron-pr-393/metron-docker/compose
130 pr-393 % docker-machine ls
NAME             ACTIVE   DRIVER       STATE     URL   SWARM   DOCKER    ERRORS
metron-machine   -        virtualbox   Stopped                 Unknown

ottobackwards commented:
I don't know enough about docker, i can exec into bash and populate zk etc so it is 'running' or something is

ottobackwards commented:
0 pr-393 %  docker ps --format 'table {{.Names}}\t{{.Status}}'

NAMES                    STATUS
metron_storm_1           Up About an hour
metron_kibana_1          Up About an hour
metron_hbase_1           Up About an hour
metron_elasticsearch_1   Up About an hour
metron_mysql_1           Up About an hour
metron_kafkazk_1         Up About an hour

merrimanr commented:
Your machine is stopped, that is the problem.  Can you go into Virtualbox and start it back up?  Then you should be able to run the eval statement.  You may have to build the environment again if you didn't build on "metron-machine".

It's likely that you are using MacOS as your Docker host which has networking limitations.  Can you try doing this using docker-machine?  My guess is that you followed the instructions in "Setup" since you do have "metron-machine", but when it came time to run "docker-compose up -d" you were in a new session and had not run eval "$(docker-machine env metron-machine)" so Docker was not pointed at the right machine.  In that case it would have used your local environment instead which will not work.

ottobackwards commented:
What step was supposed to start the virtual box machine?  How do I do this 'using docker-machine' differently than the instructions?  I followed them exactly from the top of the readme.

ottobackwards commented:
9028  checkout-pr 393
 9029  cd incubator-metron-pr-393
 9030  mvn clean install -DskipTests
 9031  export METRON_HOME=.
 9032  export METRON_DOCKER_HOME=$METRON_HOME/metron-docker
 9033  cd $METRON_DOCKER_HOME
 9034  ./scripts/create-docker-machine.sh
 9035  eval "$(docker-machine env metron-machine)"
 9036  cd compose
 9037  docker-compose up -d

merrimanr commented:
The step that starts the machine is in "Setup":

cd $METRON_DOCKER_HOME && ./scripts/create-docker-machine.sh

After you run that the machine will be created and running.  To start a stopped machine you can either go into the VirtualBox application and start it that way or you can run:

$ docker-machine start metron-machine

Or you could start over too.  Somehow you got into a bad state because your machine is not running.  That is your problem.  Maybe you are trying to do this a couple days after you initially built the environment and the machine went down between then and now?  If you want to start over, run this to remove the current metron-machine:

$ docker-machine rm metron-machine

ottobackwards commented:
So - should we destroy the machine when we are done?

ottobackwards commented:
ok - starting the machine fixed it.  I think the issue is that I had a defined, but not running machine from a previous run.  So either we need instructions to be clear on what to do with the machine when you are done... or the script maybe needs to start it if it doesn't create it?

merrimanr commented:
I usually just leave it up.  My current environment has been running for 2 weeks with no problem.  The intention behind creating a Metron-specific machine is that it won't conflict with other Docker machines someone may be using (or if people are using the default machine, which I would guess is most common).  

Not sure how yours went down but sometimes that happens.  Would adding a note in the "Setup" section of the README to make sure machine is running do it (right before the eval command)?  I could also add quick instructions on how to stop/start a machine.

ottobackwards commented:
`
  value=$(VBoxManage list vms | grep -ic "metron-machine" )
  if [ $value -eq 1 ]
  then
      VBoxManage startvm "metron-machine" --type headless                    
  fi
  
`

ottobackwards commented:
I don't think we can assume that dev's will just leave it up and running.   The flow should be close to the vagrant flow shouldn't it?

ottobackwards commented:
This does bring up an interesting point: when testing something like this - it is not enough to just run through the examples, you need to do the more realistic workflow - multiple sessions etc to catch something like this

merrimanr commented:
Fair enough.  Probably best to start/stop with each use.  I will update the docs to reflect that.  Vagrant is not ideal for me as a development tool so I don't use it as a guide :)  But I see your point and consistency is good.

Yeah I agree.  Testing with multiple sessions is a good idea and something we should do for tools like this.  In the future I will highlight that in the PR description if applicable.

ottobackwards commented:
This is a great work.  I am not sure what the right flow or scripts will end up being, but if we document it we can land it and iterate through pr's later.

ottobackwards commented:
Just waiting on the docs for a +1 on this, then we can get it in

merrimanr commented:
I added the BROKER_IP_ADDR change from @kylerichardson and now configs are loaded on container start.  Also added the various documentation requests.  @ottobackwards let me know if the start/stop documentation is clear enough.

ottobackwards commented:
+1 - this is great work Ryan

merrimanr commented:
Thanks!  Appreciate all the valuable feedback.

kylerichardson commented:
+1 Awesome contribution.

I'm already working on some follow on PRs for this :-).
